### PR TITLE
Multihead attention

### PR DESCRIPTION
**Pull request by XiaodongWuFetch**

### CODE REVIEW

File: libs/ml/tests/ml/layers/multihead_attention.cpp [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+#include "ml/layers/scaled_dot_product_attention.hpp"
+#include "ml/ops/loss_functions.hpp"
+#include "ml/optimisation/sgd_optimiser.hpp"
+#include "ml/regularisers/regulariser.hpp"
+#include "vectorise/fixed_point/fixed_point.hpp"
+#include <ml/layers/multihead_attention.hpp>
+
+#include "gtest/gtest.h"
+
+#include <chrono>
```
Comment (MissingNO57) [inline libs/ml/tests/ml/layers/multihead_attention.cpp]: Do you need all of these imports?

File: libs/ml/include/ml/layers/multihead_attention.hpp [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+#include "math/standard_functions/sqrt.hpp"
+#include "ml/layers/fully_connected.hpp"
+#include "ml/layers/scaled_dot_product_attention.hpp"
+#include "ml/ops/add.hpp"
+#include "ml/ops/concatenate.hpp"
+#include "ml/ops/divide.hpp"
+#include "ml/ops/flatten.hpp"
+#include "ml/ops/matrix_multiply.hpp"
+#include "ml/ops/placeholder.hpp"
+#include "ml/ops/transpose.hpp"
```
Comment (MissingNO57) [inline libs/ml/include/ml/layers/multihead_attention.hpp]: Do you need all of these includes? I think that we can remove at least transpose.

File: libs/ml/include/ml/layers/multihead_attention.hpp [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+    std::string query =
+        this->template AddNode<fetch::ml::ops::PlaceHolder<ArrayType>>(name + "_Query", {});
+    std::string key =
+        this->template AddNode<fetch::ml::ops::PlaceHolder<ArrayType>>(name + "_Key", {});
+    std::string value =
+        this->template AddNode<fetch::ml::ops::PlaceHolder<ArrayType>>(name + "_Value", {});
+
+    // do n_heads time linear transformation
+    std::vector<std::string> heads;
+    for (SizeType i = 0; i < n_heads_; i++)
```
Comment (MissingNO57) [inline libs/ml/include/ml/layers/multihead_attention.hpp]: Code style: SizeType i{0}

**Review by MissingNO57 — CHANGES_REQUESTED** [review]

File: libs/ml/tests/ml/layers/multihead_attention.cpp (hunk 1 above)
Comment (XiaodongWuFetch) [reply libs/ml/tests/ml/layers/multihead_attention.cpp]: nope

**Review by MissingNO57 — APPROVED** [review]
LGTM

**Pull request merged**